Implement automatic key rotation for realm keys

volck opened the pull request:
Adds scheduled task that automatically rotates realm signing and encryption keys based on configurable rotation periods. This addresses the security concern of long-lived keys and reduces operational burden.

Key features:
- Configurable rotation period per key provider
- Automatic expiration of passive keys after grace period
- Optional automatic deletion of disabled keys
- Prometheus metrics for monitoring rotation events
- Admin events for audit trail
- Support for RSA, ECDSA, HMAC, and AES key providers

The rotation task runs as part of the existing scheduled tasks infrastructure and is cluster-aware to prevent race conditions.

Closes #11693

volck committed:
Implement automatic key rotation for realm keys

Adds scheduled task that automatically rotates realm signing and
encryption keys based on configurable rotation periods. This addresses
the security concern of long-lived keys and reduces operational burden.

Key features:
- Configurable rotation period per key provider
- Automatic expiration of passive keys after grace period
- Optional automatic deletion of disabled keys
- Prometheus metrics for monitoring rotation events
- Admin events for audit trail
- Support for RSA, ECDSA, HMAC, and AES key providers

The rotation task runs as part of the existing scheduled tasks
infrastructure and is cluster-aware to prevent race conditions.

Closes #11693

Signed-off-by: Emil Volckmar Ry <emilvry@gmail.com>

ahus1 commented:
@volck - thank you for this PR. I see some emoji, it would be great if someone of those people would help to assess this PR. 

Some things I notice: 

* When setting an deletion time, keys need to be around so that they exceed the maximum idle time for any of the session types in the realm. Assuming that offline sessions could be idle for several days this should be validated. Maybe the deletion time could be derived from those settings instead leaving it to the user to configure it manually. 
* Please post some screenshots of your UIs as they are presented the user, this helps with the review. 
* There might be blocker as key rotation might not be work yet in all places. This needs at least a review - who can help? First possible issue identified: #46350

volck committed:
Derive passive key retention floor from realm session timeouts

Passive keys must remain available at least as long as the longest-lived
session type in the realm to ensure tokens signed with those keys can
still be verified. This change:

- Add computeMinimumPassiveKeyRetention() that computes max(all session
  timeouts) + safety margin (10%, min 1 hour)
- Enforce session-derived minimum in expirePassiveKeys(), logging a
  warning when the configured expiration is overridden
- Propagate deletion settings (autoDeleteDisabledKeys, deletionGracePeriod)
  to newly created providers during key rotation
- Fix duplicate imports and indentation in rotateKey()
- Add tests for session-aware retention guardrails
- Fix test cleanup to catch rotated provider names
- Fix testAutomaticKeyDeletion JPA L1 cache issue (use getComponentsStream)

Closes #46350

slaskawi reviewed (commented):
I'm much in favor of this feature! Rotating keys becomes more and more important in highly secured production environments.

The UI part looks great and the code implementation looks fine too. 

I added one question, whether or not we'd like to implement this as a Workflow?

volck commented on model/storage-private/src/main/java/org/keycloak/services/scheduled/AutomaticKeyRotationTask.java:
```diff
@@ -0,0 +1,566 @@
+ * This task checks all key providers across all realms and rotates keys that are due for rotation.
+ * 
+ */
+public class AutomaticKeyRotationTask implements ScheduledTask {
```
Thanks for taking the time to review, much appreciated. I will look into using the workflow provider if that is easier and more flexible for the enduser. I dont know if the maintainers have any preference?

slaskawi replied:
I talked with @ahus1 about this PR and the workflows recently. We agreed that workflows might be more suitable for user-facing things (e.g. expiring accounts). This however, is a core-type of functionality. So using Scheduled Task is actually OK. 

I believe this tread can be resolved.

volck replied:
I'll mark it as resolved. Thanks! 👍

volck commented:
@stianst @ahus1 anyone? :)

gh-baptiste-delafaire commented on model/storage-private/src/main/java/org/keycloak/services/scheduled/AutomaticKeyRotationTask.java:
```diff
@@ -0,0 +1,566 @@
+    /**
+     * Checks if a key should be rotated based on the last rotation time and rotation period.
+     */
+    private boolean shouldRotateKey(ComponentModel provider, RealmModel realm) {
```
`shouldRotateKey` has a hidden write side-effect
`shouldRotateKey` looks like a pure predicate, but it mutates persistent state when `lastRotationTime` is missing:

volck replied:
Fixed in the refactor (`738a053`): the init write moved out of the predicate into `initializeMissingRotationTimestamps`, so `isDueForRotation` is now a pure, side-effect-free predicate.

gh-baptiste-delafaire commented on model/storage-private/src/main/java/org/keycloak/services/scheduled/AutomaticKeyRotationTask.java:
```diff
@@ -0,0 +1,566 @@
+            MultivaluedHashMap<String, String> config = new MultivaluedHashMap<>(provider.getConfig());
+            config.putSingle(LAST_ROTATION_TIME_KEY, String.valueOf(Time.currentTimeMillis()));
+            provider.setConfig(config);
+            realm.updateComponent(provider);
```
Does `ClusterAwareScheduledTaskRunner`  protect against concurrent writes coming from other paths/pods ?  Do we have an optimistic lock here ?

volck replied:
Addressed in `738a053`: each mutation (init/rotate/expire/delete) now runs in its own short transaction that re-reads the component and re-checks the predicate before writing — the compare-and-set you suggested. The cluster runner (`executeIfNotExecuted`) already prevents concurrent task runs across nodes; this adds in-transaction CAS against concurrent admin edits. It's application-level CAS rather than a DB optimistic lock.

gh-baptiste-delafaire commented on model/storage-private/src/main/java/org/keycloak/services/scheduled/AutomaticKeyRotationTask.java:
```diff
@@ -0,0 +1,566 @@
+    private static final String DISABLED_TIME_KEY = "disabledTime";
+
+    @Override
+    public void run(KeycloakSession session) {
```
This method looks a bit weird for me as the responsabilities are not clear inside each component, IMO we should consider splitting each component, for exemple `updateLastRotationMetric` and `shouldRotateKey` both read `LAST_ROTATION_TIME_KEY` and the second one mutates it. both share the same implicit state. 

Shouldn't we consider splitting this code in small steps with one concern. 
1 - We could create a small immutable DTO :
```java
record ProviderRotationState(
        String realmId,
        String realmName,
        String providerId,
        Long lastRotationTimeMillis, // null when missing
        long rotationPeriodSeconds,
        long passiveExpirationSeconds,
        [...]
) {
    static ProviderRotationState from(ComponentModel c, RealmModel r) { parse once, with try/catch }
}
```
2 - Make the predicate and the metric pure

```java
static boolean isDueForRotation(ProviderRotationState s, long now) {
    if (!s.autoRotationEnabled() || !s.active()) return false;
    if (s.lastRotationTimeMillis() == null) return false;
    long elapsedMs = now - s.lastRotationTimeMillis();
    return elapsedMs >= TimeUnit.SECONDS.toMillis(s.rotationPeriodSeconds());
}
private void publishLastRotationGauge(ProviderRotationState s) {
    if (!s.active() || s.lastRotationTimeMillis() == null) {
        removeGauge(s);
        return;
    }
    gaugeFor(s).set(s.lastRotationTimeMillis() / 1000);
}
```
Both functions are now side-effect free with respect to persistent state. They cannot drift from each other because they read from the same snapshot.

3 - Move the "init lastRotationTime" out of the hot loop

```java
private void initializeMissingRotationTimestamps(KeycloakSessionFactory factory, RealmModel realm) {
    realm.getComponentsStream(realm.getId(), KeyProvider.class.getName())
         .filter(c -> "true".equalsIgnoreCase(c.get(AUTO_ROTATION_ENABLED_KEY)))
         .filter(c -> isBlank(c.get(LAST_ROTATION_TIME_KEY)))
         .forEach(c -> KeycloakModelUtils.runJobInTransaction(factory, txSession -> {
             RealmModel txRealm = txSession.realms().getRealm(realm.getId());
             ComponentModel fresh = txRealm.getComponent(c.getId());
             if (fresh == null || !isBlank(fresh.get(LAST_ROTATION_TIME_KEY))) return; // re-check in tx
             MultivaluedHashMap<String, String> cfg = new MultivaluedHashMap<>(fresh.getConfig());
             cfg.putSingle(LAST_ROTATION_TIME_KEY, String.valueOf(Time.currentTimeMillis()));
             fresh.setConfig(cfg);
             txRealm.updateComponent(fresh);
         }));
}
```
The re-read inside the transaction provides the compare-and-set guarantee the original code was missing. Each init is its own short transaction, so a failure on one provider doesn't roll back the others.

4 - The new run() becomes a clean pipeline

```java
public void run(KeycloakSession session) {
    long now = Time.currentTimeMillis();
    long start = now;
    session.realms().getRealmsStream().forEach(realm -> {
        try {
            initializeMissingRotationTimestamps(session.getKeycloakSessionFactory(), realm);
            List<ProviderRotationState> snapshots = realm
                    .getComponentsStream(realm.getId(), KeyProvider.class.getName())
                    .map(c -> ProviderRotationState.from(c, realm))
                    .toList();
            // Pure: publish observability from the snapshot
            snapshots.forEach(this::publishLastRotationGauge);
            // Decide, then act. Each action runs in its own short transaction.
            for (ProviderRotationState s : snapshots) {
                if (s.autoRotationEnabled() && isDueForRotation(s, now)) {
                    rotateKeyTx(session.getKeycloakSessionFactory(), s);
                }
                if (!s.active() && shouldExpire(s, now, realm)) {
                    expirePassiveKeyTx(session.getKeycloakSessionFactory(), s);
                }
                if (shouldDelete(s, now)) {
                    deleteDisabledKeyTx(session.getKeycloakSessionFactory(), s);
                }
            }
        } catch (Exception e) {
            logger.errorv(e, "Automatic key rotation failed for realm '{0}'", realm.getName());
        }
    });
    log(start);
```

So the improvements are : 
- single source of truth per iteration, metric and decision both read identical data.
- no write inside a predicate
- Init, rotation, expiration and deletion each own their transaction, with re-check inside, removing the silent last-write-wins issue and shortening lock duration.
- A bad provider (e.g. malformed lastRotationTime) fails its own transaction and is logged; it doesn't poison the entire realm scan.
- The orchestration in run() is ~20 lines and reads as a sentence: snapshot → publish → decide → act.

volck replied:
Done in `738a053`: introduced an immutable `ProviderRotationState` parsed once via `from()`. The metric (`publishLastRotationGauge`) and the decision (`isDueForRotation`) now read from the same snapshot so they can't drift, and `run()` reads as snapshot → publish → decide → act.

gh-baptiste-delafaire reviewed (approved):
That a very important and very nice improvement, I can't wait to see this in Keycloak ! 
THANKS 🙏

gh-baptiste-delafaire commented:
Hi guys, do we have any ETA about a possible release ? It will be a very great change for keycloak

volck commented:
Thanks for your feedback @gh-baptiste-delafaire, very much appreciated. I'd like to address them but I'm a bit unsure what to do about it given maintainers have not said what they want. Once they do, I'll happily address them.

eicki commented:
Is there a time window in which a new signing key will be available in passive state before rotation happens, some kind of pre-activation phase?
If clients are smart enough to fetch keys from the jwks endpoint on demand that is no issue, but if clients only fetch keys based on a cron job or even manual action is required, a pre-activation phase is needed.

volck commented:
> Is there a time window in which a new signing key will be available in passive state before rotation happens, some kind of pre-activation phase? If clients are smart enough to fetch keys from the jwks endpoint on demand that is no issue, but if clients only fetch keys based on a cron job or even manual action is required, a pre-activation phase is needed.

Seems like a great addition. The timewindow should probably be configurable? or should it be derived based on something else?

volck committed:
refactor(services): make key rotation task side-effect-free and transaction-safe

Address review feedback on the automatic key rotation task:

- Introduce immutable ProviderRotationState snapshot parsed once per
  provider, so metric and decision logic share one source of truth.
- Make isDueForRotation and passive-key predicates pure; remove the
  hidden write side-effect from the former shouldRotateKey.
- Run each mutation (init, rotate, expire, delete) in its own short
  transaction with an in-tx re-check, giving compare-and-set semantics
  and isolating failures to a single provider.

volck commented:
> Thanks for your feedback @gh-baptiste-delafaire, very much appreciated. I'd like to address them but I'm a bit unsure what to do about it given maintainers have not said what they want. Once they do, I'll happily address them.

got bored of waiting and (had some time) so I decided to try to address the issues you pointed out @gh-baptiste-delafaire, see if these makes sense. I think having separated concerns makes a lot of sense here.

volck committed:
fix(services): preserve key params and harden automatic key rotation

Address Copilot review findings on the automatic key rotation task:

- Preserve all key-generation parameters on rotation via an allowlist
  (secret size, EC curve, certificate generation, ...), so AES/HMAC and
  EC keys are no longer silently downgraded to defaults.
- Isolate each provider's rotation/expiration/deletion (and timestamp
  init) in its own try/catch so one failure no longer skips the rest of
  the realm.
- Treat a non-positive rotationPeriod as invalid and fall back to the
  90-day default instead of rotating on every run.
- Never delete key material when deletionGracePeriod is negative.
- Fix a MessageFormat log placeholder (%s -> {0}).

Adds regression tests covering HMAC secret size, ECDSA curve, negative
deletion grace period, and zero rotation period.

volck committed:
fix(services): do not rotate administratively disabled key providers

A provider can be active=true while enabled=false (administratively
disabled). isDueForRotation previously ignored the enabled flag, so such
a provider was rotated into an enabled replacement, silently re-enabling
key material an administrator had turned off.

Require enabled=true for rotation and add a regression test.

volck committed:
fix(services): address remaining key rotation review findings

- Include access-token and action-token lifespans in the passive-key
  retention floor so a key is not disabled while such tokens can still
  need verification.
- Dispatch rotation admin events to the realm's configured and global
  EventListenerProviders (not just the event store), honoring
  isAdminEventsDetailsEnabled, so audit/SIEM listeners observe them.
- Reconcile the last-rotated gauges each scan and remove meters for
  providers or realms deleted administratively, preventing unbounded
  stale metric cardinality.
- Tighten testKeyRotationNotDueYet to detect an unexpected rotation
  replacement.

Adds regression tests for the action-token retention floor and gauge
cleanup after provider deletion.

volck committed:
test(keys): strengthen action-token retention test and add admin-event test

- Fix testPassiveRetentionIncludesActionTokenLifespan: the passive key
  state was not being persisted, so it never exercised the expiration
  path (a false green). Set the passive/expiration config via a committed
  update transaction, verified red without the retention-floor fix.
- Add testRotationFiresAdminEvent using @InjectAdminEvents to verify a
  rotation dispatches a COMPONENT admin event through the event pipeline.

volck commented on model/storage-private/src/main/java/org/keycloak/services/scheduled/AutomaticKeyRotationTask.java:
```diff
@@ -0,0 +1,700 @@
+        // Copy key size if present
+        if (currentProvider.contains(KEY_SIZE_KEY)) {
+            config.putSingle(KEY_SIZE_KEY, currentProvider.get(KEY_SIZE_KEY));
+        }
```
Fixed in `945322b`: rotation now copies all generation parameters via a `GENERATION_PARAM_KEYS` allowlist (`secretSize`, `ecdsa`/`ecdh`/`eddsaEllipticCurveKey`, `ecGenerateCertificate`, …). Added regression tests for HMAC secret size and ECDSA curve (P-521 no longer downgrades to P-256).

volck commented on model/storage-private/src/main/java/org/keycloak/services/scheduled/AutomaticKeyRotationTask.java:
```diff
@@ -0,0 +1,700 @@
+                    .filter(state -> state.autoRotationEnabled() && state.lastRotationTimeMillis() == null)
+                    .forEach(state -> initializeMissingRotationTimestamps(factory, state));
+
+            // Publish observability from the snapshot. Pure with respect to persistent state.
+            snapshots.forEach(this::publishLastRotationGauge);
+
+            counters.autoRotationEnabled += (int) snapshots.stream()
+                    .filter(ProviderRotationState::autoRotationEnabled)
+                    .count();
+
+            // Decide, then act. Each action re-reads and re-checks inside its own transaction.
+            for (ProviderRotationState state : snapshots) {
+                if (isDueForRotation(state, now) && rotateKeyTx(factory, state, now)) {
+                    counters.rotated++;
+                }
+                if (shouldExpirePassiveKey(state, now) && expirePassiveKeyTx(factory, state, now)) {
+                    counters.expired++;
+                }
+                if (shouldDeleteDisabledKey(state, now) && deleteDisabledKeyTx(factory, state, now)) {
+                    counters.deleted++;
+                }
```
Fixed in `945322b`: each provider's actions — and the timestamp init — now run inside their own `try/catch` in `processRealm`, so one provider's failure is logged and no longer skips the rest of the realm.

volck commented on services/src/main/java/org/keycloak/keys/Attributes.java:
```diff
@@ -92,4 +92,27 @@ public interface Attributes {
+    String AUTO_ROTATION_ENABLED_KEY = "autoRotationEnabled";
+    ProviderConfigProperty AUTO_ROTATION_ENABLED_PROPERTY = new ProviderConfigProperty(AUTO_ROTATION_ENABLED_KEY, "Enable Automatic Rotation", "Enable automatic key rotation for this provider", BOOLEAN_TYPE, "false");
+
+    String ROTATION_PERIOD_KEY = "rotationPeriod";
+    ProviderConfigProperty ROTATION_PERIOD_PROPERTY = new ProviderConfigProperty(ROTATION_PERIOD_KEY, "Rotation Period (seconds)", "Number of seconds between automatic key rotations", STRING_TYPE, "7776000");
+
+    String PASSIVE_KEY_EXPIRATION_KEY = "passiveKeyExpiration";
+    ProviderConfigProperty PASSIVE_KEY_EXPIRATION_PROPERTY = new ProviderConfigProperty(PASSIVE_KEY_EXPIRATION_KEY, "Passive Key Expiration (seconds)", "Number of seconds to keep passive keys before they are disabled", STRING_TYPE, "2592000");
```
Hardened at runtime in `945322b`: a non-positive `rotationPeriod` now falls back to the 90-day default (no more rotate-every-run), and a malformed value causes the provider to be skipped (logged) rather than mis-rotated. I kept this in the task rather than adding factory-level input validation — happy to add `validateConfiguration` checks to the key-provider factories as well if you'd prefer to reject bad values on create/update.

volck commented on services/src/main/java/org/keycloak/keys/Attributes.java:
```diff
@@ -92,4 +92,27 @@ public interface Attributes {
+    String AUTO_DELETE_DISABLED_KEYS_KEY = "autoDeleteDisabledKeys";
+    ProviderConfigProperty AUTO_DELETE_DISABLED_KEYS_PROPERTY = new ProviderConfigProperty(AUTO_DELETE_DISABLED_KEYS_KEY, "Auto-Delete Disabled Keys", "Automatically delete disabled keys after grace period expires", BOOLEAN_TYPE, "false");
+
+    String DELETION_GRACE_PERIOD_KEY = "deletionGracePeriod";
+    ProviderConfigProperty DELETION_GRACE_PERIOD_PROPERTY = new ProviderConfigProperty(DELETION_GRACE_PERIOD_KEY, "Deletion Grace Period (seconds)", "Number of seconds to wait after disabling a key before deleting it (default 3600 = 1 hour)", STRING_TYPE, "3600");
```
The data-loss path is closed in `945322b`: a negative `deletionGracePeriod` never triggers deletion (`shouldDeleteDisabledKey` returns false), with a regression test. As with the rotation period, this is a runtime guard rather than factory-level validation — I can add the latter too if preferred.

volck commented on model/storage-private/src/main/java/org/keycloak/services/scheduled/AutomaticKeyRotationTask.java:
```diff
@@ -0,0 +1,700 @@
+    static long computeMinimumPassiveKeyRetention(RealmModel realm) {
+        long maxLifespanSeconds = 0;
+
+        // SSO session lifespans (in seconds)
+        maxLifespanSeconds = Math.max(maxLifespanSeconds, realm.getSsoSessionMaxLifespan());
+        maxLifespanSeconds = Math.max(maxLifespanSeconds, realm.getSsoSessionIdleTimeout());
+
+        // "Remember me" variants (0 means "use regular SSO value", so skip if 0)
+        if (realm.getSsoSessionMaxLifespanRememberMe() > 0) {
+            maxLifespanSeconds = Math.max(maxLifespanSeconds, realm.getSsoSessionMaxLifespanRememberMe());
+        }
+        if (realm.getSsoSessionIdleTimeoutRememberMe() > 0) {
+            maxLifespanSeconds = Math.max(maxLifespanSeconds, realm.getSsoSessionIdleTimeoutRememberMe());
+        }
+
+        // Offline session idle timeout — typically the largest (default 30 days)
+        maxLifespanSeconds = Math.max(maxLifespanSeconds, realm.getOfflineSessionIdleTimeout());
```
Addressed in `9dffd35`: `computeMinimumPassiveKeyRetention` now also includes access-token and action-token lifespans (admin + user, including per-action overrides), not just session timeouts. Added a test that a passive key isn't expired while the action-token floor still covers it.

volck commented on model/storage-private/src/main/java/org/keycloak/services/scheduled/AutomaticKeyRotationTask.java:
```diff
@@ -0,0 +1,700 @@
+                }
+                
+                // Always save with full details for system-generated security events
+                eventStore.onEvent(adminEvent, true);
```
Fixed in `9dffd35`: `fireAdminEvent` now dispatches to the realm's configured and global `EventListenerProvider`s in addition to the store, and honors `isAdminEventsDetailsEnabled` — mirroring the normal admin-event path. Added a test using `@InjectAdminEvents`.

volck commented on tests/base/src/test/java/org/keycloak/tests/keys/AutomaticKeyRotationTest.java:
```diff
@@ -0,0 +1,671 @@
+                .query(realm.getId(), KeyProvider.class.getName())
+                .stream()
+                .filter(p -> "rsa-generated".equals(p.getProviderId()))
+                .filter(p -> p.getName().startsWith(keyProvider.getName())) // Check for our provider or rotated versions
```
Tightened in `945322b`/`aec3269`: `testKeyRotationNotDueYet` now explicitly asserts no `rsa-generated-<timestamp>` replacement was created, so an unexpected rotation is detected.

volck commented on model/storage-private/src/main/java/org/keycloak/services/scheduled/AutomaticKeyRotationTask.java:
```diff
@@ -0,0 +1,700 @@
+                Gauge.builder(KEY_LAST_ROTATED_METER_NAME, value, AtomicLong::get)
+                    .description("Unix timestamp (seconds) when key was last rotated")
+                    .tags(tags)
+                    .register(Metrics.globalRegistry);
```
Fixed in `9dffd35`: gauges are tracked with their meter (`GaugeHandle`) and a `reconcileStaleGauges` step each scan removes meters for providers/realms that no longer exist, preventing unbounded stale cardinality. Added a test that deletes a provider and asserts the meter is removed.

volck commented on model/storage-private/src/main/java/org/keycloak/services/scheduled/AutomaticKeyRotationTask.java:
```diff
@@ -0,0 +1,700 @@
+                eventStore.onEvent(adminEvent, true);
+            }
+        } catch (Exception e) {
+            logger.warnv(e, "Failed to fire admin event for key rotation in realm '%s'", realm.getName());
```
Fixed in `945322b`: changed the `warnv` placeholder from `%s` to `{0}` so the realm name is included.

volck commented on model/storage-private/src/main/java/org/keycloak/services/scheduled/AutomaticKeyRotationTask.java:
```diff
@@ -0,0 +1,735 @@
+    /**
+     * Pure predicate: is an active, auto-rotation-enabled key past its rotation period?
+     */
+    static boolean isDueForRotation(ProviderRotationState state, long now) {
+        if (!state.autoRotationEnabled() || !state.active()) {
+            return false;
+        }
```
Fixed in `702f052`: `isDueForRotation` now also requires `enabled`, so an administratively disabled (`enabled=false`) provider is never rotated and can't be silently re-enabled. Added a regression test.

volck committed:
fix(services): harden automatic key rotation per review 4839567779

Address four items from the Copilot re-review:

A. Passive-key retention floor now also accounts for per-client
   access-token/session lifespan overrides and realm client-default
   lifespans, so a key is not expired while a client token it signed is
   still valid.
B. A disabled provider with auto-delete enabled but no disabledTime (the
   normal state after a manual disable, or enabling auto-delete on an
   already-disabled provider) now has its disable timestamp initialized
   transactionally so the deletion grace period can start, instead of
   being retained forever.
C. Preserve the ecdhAlgorithm generation parameter on rotation so an
   ECDH provider with a non-default key-agreement algorithm is not
   silently reset to ECDH-ES.
D. Gate counter/gauge creation, publication and reconciliation on
   metrics-enabled, matching ExpirationTaskFactory and
   PasswordCredentialProviderFactory, so installations with metrics
   disabled do not pay the memory/cardinality cost.

test(keys): add regression tests for A-D; enable metrics on the test
server via ServerConfigWithMetrics so the gauge/counter assertions run.

volck committed:
feat(services): add optional pre-activation phase to automatic key rotation

Address eicki's review request: give clients that cache the JWKS (fetch
on a cron or manual trigger) time to pick up a new signing key before it
is used.

A new key provider config `preActivationPeriod` (seconds, default 0)
controls this. When 0, rotation activates the replacement key immediately
(unchanged behavior). When greater than 0, a due rotation instead
publishes the replacement key as passive-but-enabled (visible in the JWKS
but not signing) with an `activationTime` in the future, while the
current key keeps signing. The current key is marked `preActivationPending`
so it is not rotated again while a promotion is pending. Once the
activation time passes, a later run promotes the new key to active and
demotes its predecessor to passive, starting the predecessor's normal
passive-expiration lifecycle.

Rotation now reports a `promoted` counter and emits `pre-activate`/
`promote` metric operations; both phases run in their own transaction with
a re-check for multi-node safety.

test(keys): add regression tests for the passive-publish, promotion, and
no-duplicate-while-pending behaviors.

gh-baptiste-delafaire commented:
Thanks a lot, I hope that owner will approve this pr very soon ! looking foward to see it in production

slaskawi commented on model/storage-private/src/main/java/org/keycloak/services/scheduled/AutomaticKeyRotationTask.java:
```diff
@@ -0,0 +1,1129 @@
+            long minimumPassiveRetentionSeconds = computeMinimumPassiveKeyRetention(realm);
+
+            // Snapshot: parse each provider's configuration exactly once into an immutable state object.
+            List<ProviderRotationState> snapshots = realm
```
This approach seems sub-optimal. We collect all the snapshots before checking if the auto-rotation is turned on. I think it would be better to flip the logic here - first check if it's turned on and **then** collect the snapshots.

slaskawi commented on model/storage-private/src/main/java/org/keycloak/services/scheduled/AutomaticKeyRotationTask.java:
```diff
@@ -0,0 +1,1129 @@
+    }
+
+    // -------------------------------------------------------------------------
+    // Immutable snapshot + pure decision helpers
```
nit: Personally, I'm not a big fan of these comments. They strongly suggest that something should go to another file but for some reason didn't. Could we either remove it or ensure this class is split properly?

slaskawi commented on model/storage-private/src/main/java/org/keycloak/services/scheduled/AutomaticKeyRotationTask.java:
```diff
@@ -0,0 +1,1129 @@
+        if (!state.autoRotationEnabled() || !state.active() || !state.enabled()) {
+            return false;
+        }
+        if (state.activationTimeMillis() != null || state.preActivationPending()) {
```
Pre-activation marks the active predecessor as pending. If an administrator deletes the pending replacement before activation, no task path clears that predecessor flag, so this predicate skips the still-active predecessor forever.

slaskawi commented on model/storage-private/src/main/java/org/keycloak/services/scheduled/AutomaticKeyRotationTask.java:
```diff
@@ -0,0 +1,1129 @@
+            if (component == null) {
+                return false;
+            }
+            ProviderRotationState fresh = ProviderRotationState.from(component, realm, computeMinimumPassiveKeyRetention(realm));
```
The task lock expires after the configured interval and is not renewed. If this run exceeds that TTL, two nodes can both pass this re-check against the same due provider and each add a replacement. Please add provider-level serialization (or a renewed lease) and cover an overlapping run.

slaskawi reviewed (commented):
I'd be definitely in favor of getting this feature into the main codebase! Automatic Realm Key rotation is a very robust method for satisfying various compliance requirements. 

The code looks generally OK to me (though I found a few things that could be improved). 

The biggest missing piece at the moment is documentation. We'd definitely need put information about this feature into the Server Admin Guide as well as Release Notes.

mabartos commented:
@sguilhen @pedroigor Workflows feature could be reused for this, right?

volck commented:
> @sguilhen @pedroigor Workflows feature could be reused for this, right?

according to https://github.com/keycloak/keycloak/pull/45091#discussion_r3137089133 this seems to conclude that it might be a scheduled task.
